### PR TITLE
Refactor LBSProblem data member access and improve factory construction.

### PR DESCRIPTION
**Pull request by wdhawkins**
This PR make the following improvements to the problem classes:

1. Remove friend-based restart I/O access.
2. Eliminate the problem initialization flag.
3. Improve factory-built runtime construction.
4. Clarify public versus internal solver APIs. 
5. Split outflow balance from cell transport metadata and make transport views const.

## PR Checklist

- [x] I have updated the user guide and/or Python API documentation if necessary.

### CODE REVIEW

**Review by andrsd — APPROVED** [review]

File: modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc [hunk 1]
```diff
@@ -613,10 +613,28 @@ DiscreteOrdinatesProblem::SetSweepChunkMode(SweepChunkMode mode)
...
-  if (initialized_)
+  if (discretization_)
     UpdateAngularFluxStorage();
 }
 
+void
+DiscreteOrdinatesProblem::RebuildBoundaryRuntimeData()
+{
+  if (not grid_face_histogram_)
+    return;
```
Comment (quocdang1998) [inline modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc]: Do we have to rebuild `FluxDataStructures` too?

File: modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc [hunk 2]
```diff
@@ -613,10 +613,28 @@ DiscreteOrdinatesProblem::SetSweepChunkMode(SweepChunkMode mode)
...
   sweep_chunk_mode_ = mode;
-  if (initialized_)
+  if (discretization_)
     UpdateAngularFluxStorage();
 }
 
+void
+DiscreteOrdinatesProblem::RebuildBoundaryRuntimeData()
+{
+  if (not grid_face_histogram_)
```
Comment (quocdang1998) [inline modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc]: Why do we skip if `grid_face_histogram_` was not initialized?

File: modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc [hunk 3]
```diff
@@ -345,12 +345,14 @@ DiscreteOrdinatesProblem::SetBoundaryOptions(const InputParameters& params)
   }
   const auto bid = it->second;
   boundary_definitions_[bid] = CreateBoundaryFromParams(params);
+  RebuildBoundaryRuntimeData();
```
Comment (quocdang1998) [inline modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc]: `SetBoundaryOptions` is called in 2 places: either during constructor of `DiscreteOrdinatesProblem` or during runtime through `SetBoundaryOptions`.

I think we should find a better way to avoid clean/init/initialize solver scheme for every boundary entry in the constructor. What do you think?

File: modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc [hunk 4]
```diff
@@ -613,10 +613,28 @@ DiscreteOrdinatesProblem::SetSweepChunkMode(SweepChunkMode mode)
...
     UpdateAngularFluxStorage();
 }
 
+void
+DiscreteOrdinatesProblem::RebuildBoundaryRuntimeData()
+{
+  if (not grid_face_histogram_)
+    return;
+
+  InitializeBoundaries();
```
Comment (quocdang1998) [inline modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc:652]: Don't forget to update the boundary map of the new objects to  `AngleAggregation`, `AAHD_FLUDSCommonData` and possibly `AngleSet`.

**Review by quocdang1998 — CHANGES_REQUESTED** [review]

File: modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.h [hunk 5]
```diff
@@ -50,20 +45,11 @@ class DiscreteOrdinatesProblem : public LBSProblem
...
-  void ResetMode(SweepChunkMode target_mode);
-
-  /// Rebuild WGS/AGS solver schemes (e.g., after changing sweep chunk mode).
-  void ReinitializeSolverSchemes();
-
   ~DiscreteOrdinatesProblem() override;
 
   using BoundaryDefinition = std::pair<LBSBoundaryType, std::shared_ptr<SweepBoundary>>;
 
+  /// Problem metadata and solver access.
```
Comment (quocdang1998) [inline modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.h]: Is this comment misplaced? `sweep_type_` is either AAH or CBC.

**Review by quocdang1998 — CHANGES_REQUESTED** [review]

File: modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc (hunk 1 above)
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc]: Fixed. I've modified the boundary update routine to update all of the needed data structures. I've also included a new regression test that tests a number of boundary mutations after construction.

File: modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc (hunk 2 above)
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc]: Fixed. This was a poor choice for determining runtime build status.

File: modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc (hunk 3 above)
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc]: Fixed.

File: modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc (hunk 4 above)
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.cc:652]: Fixed.

File: modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.h (hunk 5 above)
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.h]: No, it's a header line indicating the purpose of the methods that follow. I'll make them more explicit.

Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.h]: Fixed. Hopefully they read now more as section dividers instead of comments. I can always remove them as well.

File: modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.h [hunk 6]
```diff
@@ -50,20 +45,11 @@ class DiscreteOrdinatesProblem : public LBSProblem
...
-  void ResetMode(SweepChunkMode target_mode);
-
-  /// Rebuild WGS/AGS solver schemes (e.g., after changing sweep chunk mode).
-  void ReinitializeSolverSchemes();
-
   ~DiscreteOrdinatesProblem() override;
 
   using BoundaryDefinition = std::pair<LBSBoundaryType, std::shared_ptr<SweepBoundary>>;
 
+  // -- Problem metadata and solver access --
```
Comment (quocdang1998) [inline modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.h]: If you want to mark section divider, you should use Doxygen style section divider like:

```
/// \name .....
/// \{
....
/// \}
```

See `modules/linear_boltzmann_solvers/discrete_ordinates_problem/sweep/fluds/aahd_fluds.h` for example.

Furthermore, can you rebase this PR on top of the latest PR of main?

**Review by quocdang1998 — CHANGES_REQUESTED** [review]

File: modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.h (hunk 6 above)
Comment (wdhawkins) [reply modules/linear_boltzmann_solvers/discrete_ordinates_problem/discrete_ordinates_problem.h]: Fixed and rebased.

**Review by quocdang1998 — APPROVED** [review]
